sw/mit/eda/tsp/tsp.1
.\" -*- nroff -*-
.\" (c) Copyright 2005, Staal Vinterbo, all rights reserved.
.\"
.\" This file is part of TSP.
.\"
.\" TSP is free software; you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation; either version 2 of the License, or
.\" (at your option) any later version.
.\"
.\" TSP is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" along with TSP; if not, write to the Free Software
.\" Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
.\"
.TH TSP 1 TSP
.SH NAME
tsp \- compute a schedule for the traveling salesman
.SH SYNOPSIS
.B tsp 
[
.I options
]
.SH AVAILABILITY
The tsp program should be considered as being alpha testing
software. It is supplied as is, and NO WARRANTY whatsoever is given.

TSP is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License,
.HTML <a href=http://www.gnu.org/copyleft/gpl.html> (here)</a>,
or (at your option) any later version.


TSP is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with TSP; if not, write to the Free Software
Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

.HTML Get a gzipped tar archive <a href=tsp.tgz>here</a>.

.SH DESCRIPTION
.B tsp 
is a utility that searches for tours using a configurable
estimation of distributions algorithm. The input is a square matrix
representation of the distances between the points that the salesman
has to visit exactly once. Any single precision floating point values
are acceptable. The value of 1000000 is interpreted as a
non-connection.
.LP
The output is a permutation of the natural numbers 0 through number of
cities - 1, representing the tour.
.SH OPTIONS
.TP
.B -v, --verbose
Spew information...
.TP
.B -h, --help
print rudimentary help on usage.
.TP
.B -p, --psize  <value>
Default value: 100.
Population sise to use
.TP
.B --dv  <value>
Default value: 10000.
Sets dv used in probability scaling.
.TP
.B --df  <value>
Default value: 100.
Sets df used in probability scaling.
.TP
.B -n, --niter  <value>
Default value: 100.
Sets number of generations with
no improvement to wait before quitting.
.TP
.B -e, --nest  <value>
Default value: 30.
Sets number of best individuals to use each
generation to update the distribution estimate.
.TP
.B -f, --logfile  <value>
Default value: cerr.
Where to print log.
.TP
.B -l, --log
Turn on logging of generational stats.
.TP
.B --loginterval  <value>
Default value: 20.
Sets the number of generations to wait between each log output.
.TP
.B -a, --alpha  <value>
Default value: 0.15.
Sets alpha, the distribution update speed.
After each generation a subsample is used to estimate the
distribution, initially this is uniform. Then, the distribution is
updated using the previous distribution estimate and this current
estimate. 
.I alpha 
governs the speed of this according to
.EQ
next = (1-alpha)*old + alpha*current.
.EN
.TP
.B -i, --nelite  <value>
Default value: 3.
How many of the best to transfer to the next population.
.TP
.B -s, --seed  <value>
Default value: none.
Initialize random number generator with the given seed.
.TP
.B -m, --maxgen  <value>
Default value: none.
Maximal number of generations to run.
.TP
.B --version
Default value: none.
Print version and exit.
.TP
.B -r, --report
Default value: none.
Also report number of generations and score.
.SH INPUT FORMATS
.SS distance matrix
The square 
.I n 
times 
.I n 
distance matrix must be input as 
.I n
lines of text containing exactly
.I n
numbers, either single
precision floating point numbers, or integers that can fit into a
single point floating point number.
.SH EXAMPLES
$ tsp < intputfile
.P
$ cat inputfile | tsp -l --loginterval 10 --dv 1000 --alpha 0.6 -p 500 
.P
$ tsp -h
.LP
If the input file 
.I example.input 
contains:
.P
0 6.231753 6.417581
.br
1e+06 0 9.933639
.br
5.936391 7.555658 0
.P
Then we can expect output of
.br

.br
$ tsp -l -n 10 -v < example.input
.br

.br
to resemble something like:
.br

.br
Parameter verbose ('v') was set to 1 
.br
Parameter niter ('n') was set to 10 
.br
Parameter log ('l') was set to 1 
.br
EDA: psize 100 nelite: 3 niter: 10 nest: 30 
.br
Evaluating initial pop... 
.br
done. 
.br
eda.run()... 
.br
Starting evolution... 
.br
0 22.1018 250020 1.00001e+06 22.1018 
.br
20 22.1018 22.1018 22.1018 22.1018 
.br
done. 
.br
generations: 25 best: 22.1018 
.br
1 2 0
.P
(Note that from the output above, only the last line is on stdout. The
rest is on stderr).
.SH SEE ALSO
eda(3)
.SH AUTHOR
Staal Vinterbo <staal@dsg.harvard.edu>
.SH COPYRIGHT
Copyright (C) 2005 Staal Vinterbo
.SH DEPENDENCIES
This program is dependent on the eda and GNU gmp libraries, and has only been
tested  with g++ 3.3.{1,3}.
.SH FEATURES
The program requires the input from stdin. Even though the value
1000000 (one million) is interpreted as a non-connection, there is no
guarantee that a tour will only consist of cities between there is a
different cost. 
.SH BUGS
Must be plenty. Please report them to the
author. Thank you.





sw/mit/eda/tsp/tsp.cc
/* -*-C++-*-
********************************************************************************
*
* File:         tsp.cc
* RCS:          $Header: $
* Description:  
* Author:       Staal Vinterbo
* Created:      Tue Apr 19 10:45:15 2005
* Modified:     Wed May 11 15:37:08 2005 (Staal Vinterbo) staal@MGV
* Language:     C++
* Package:      TSP
* Status:       Experimental
*
* (c) Copyright 2005, Staal Vinterbo, all rights reserved.
*
* This file is part of TSP.
*
* TSP is free software; you can redistribute it and/or modify
* it under the terms of the GNU General Public License as published by
* the Free Software Foundation; either version 2 of the License, or
* (at your option) any later version.
*
* TSP is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
* GNU General Public License for more details.
*
* You should have received a copy of the GNU General Public License
* along with TSP; if not, write to the Free Software
* Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
*
********************************************************************************
*/

#include "tsp.h"
#include <string>
#include <iostream>
#include <sstream>
#include <vector>
#include <algorithm>
#include <iterator>
#include <assert.h>
#include <functional>

using std::string;
using std::istringstream;
using std::istream;
using std::ostream;
using std::vector;
using std::min;
using std::max;
using std::istream_iterator;
using std::cerr;
using std::max_element;
using std::equal_to;


template <typename T>
ostream& operator<<(ostream &stream, vector<T> &seq)
 {
   stream << "[";
   for(typename vector<T>::iterator i = seq.begin();
       i != seq.end(); )
   {
     stream << *i;
     i++;
     if(i != seq.end())
       stream << " ";
   }
   stream << "]\n";
   return stream;
 }

istream & tsp::read(istream& stream)
{
  matrix.clear();
  string buffer;
  size_t maxlen = 0;
  size_t minlen = barrier;

  while(getline(stream, buffer)){
    istringstream iss(buffer);
    istream_iterator<dist_t> iis(iss);
    vector<dist_t> v(iis,istream_iterator<dist_t>());

    size_t vsize = v.size();

    // figure out the largest value not equal to barrier...
    vector<dist_t> cop;

    remove_copy_if(v.begin(), v.end(), 
               back_inserter(cop),
               bind2nd(equal_to<dist_t>(), (dist_t)barrier));

    maxstep = max(maxstep, *max_element(cop.begin(), cop.end()));

    maxlen = max(maxlen, vsize);
    minlen = min(minlen, vsize);

    matrix.push_back(v);
  }

  if((minlen != maxlen) || (maxlen != matrix.size())){
    cerr << "Matrix not square, bye.\n";
    exit(-1);
  }
  size = minlen;

  // compute the holes...
  for(size_t i = 0; i < size; i++)
    for(size_t j = 0; j < size; j++)
      if(matrix[i][j] == (dist_t)barrier)
        barriers.push_back(pair<int,int>(i,j));
  
  assert(maxstep > static_cast<dist_t>(0.0));

  //  cerr << "Read Matrix: \n" << matrix;

  return stream;
}

dist_t tsp::length(vector<int>& path)
{

  vector<int>::iterator i = path.begin();
  if(i == path.end()){
    throw string("tsp::length, empty path supplied");
  }
  
  dist_t sum = 0.0;
  int from = *i;

  while(++i != path.end()){
    int to = *i;
    assert((size_t)from < size && (size_t)to < size);
    sum += matrix[from][to];
    from = to;
  }

  sum += matrix[from][path[0]];
  return sum;
}

